Add Bookstore example (#143)
vaticle/typedb-examples: commerce/bookstore/README.md

# Bookstore example

[![forthebadge](https://forthebadge.com/images/badges/made-with-python.svg)](http://forthebadge.com) [![forthebadge](images/made-with-typedb.svg)](https://forthebadge.com)

This is a simple example of using TypeDB.

## Introduction

> We have imaginary online Bookstore and our script/application implements some very basic functions:

Reviewer: ```suggestion
We have an imaginary online bookstore and our script/application implements some very basic functions:
```

Author: Fixed this in the commit https://github.com/vaticle/typedb-examples/pull/143/commits/75040e894ff72e634eb3b9d5456272183e35d897

- Search for a Book
- Search for a User
- Search for an Order

> - Search for books by genre

Reviewer: ```suggestion
- Search for a book
- Search for a user
- Search for an order
- Search for books by genre
```
Let's use consistent casing. It's actually quite common to refer to `Book`, `User` and so on when they are SQL tables, but for TypeDB that doesn't really hold.

Author: But the entities are written with the capital first letter. As they are defined in the schema.tql

Reviewer: We usually use `kebab-case` for schema objects in TypeDB - so we can kill two birds with one stone here: update the schema, and update the docs.

Author: kebab case implemented in https://github.com/vaticle/typedb-examples/pull/143/commits/cf7a8c3505cb93fdbdf9a96851da9b52b93715c4 
lower case for schema elements implemented in https://github.com/vaticle/typedb-examples/pull/143/commits/bac11c1f22a9e32656a3a486261a39c8a21646c0

This application consists of two python scripts. It has not much in terms of usability as it is merely a demonstration of TypeDB queries. 

> You can clearly see the implementation of mentioned functions in the source codes.

Reviewer: I think the wording here is a little bit shaky, and we can do better by adding more emphasis to the fact that the code follows a structure where each of the aforementioned functions is cleanly separated. Something like:
```suggestion
Each of the functions listed above is implemented in its own Python method.
```
It's not a very significant statement, as this would be the expectation anyway, so I would attach it to the end of the previous paragraph, instead of creating a new paragraph. You might also consider linking the file(s) where the functions are implemented.

Author: My initial version clearly (I hope) states that there is an ability to go there and see the source codes for each (of the 4 previously mentioned) function. It is obvious, but the idea itself is to motivate a newbie to see how it is done in the example. 

The proposed version just states that functions are implemented in separate methods. This is great, but the end goal is not separate methods/functions. It is to send a user to source codes to **see** the implementation themselves. Sorry, I'm not seeing how this change would make things better in regard to this goal. Let's discuss that, maybe I'm not seeing something.

Author: Also, there are only two files there. And they are clearly described later in this readme. I'm not sure linking one of the two files here would do any good. We want users to go see themselves after all.

Reviewer: Ok, how about changing the last line to:
```suggestion
You can find the implementation of each function listed above in the source code.
```

Author: Sounds good!

## Prerequisites

* [TypeDB](https://docs.vaticle.com/docs/running-typedb/install-and-run) v2.14.1+
* Python v.3.9+
* `typedb.client` — [python client](https://docs.vaticle.com/docs/client-api/python) for TypeDB

> * This repository

Reviewer: Love it. Yes, technically you do need to have this repository in order to do anything with it. And to be reading this README 😛

## Quickstart

1. Checkout this repository: `git clone https://github.com/vaticle/typedb-examples && cd typedb-examples`
2. Start the [TypeDB Server](http://docs.vaticle.com/docs/running-typedb/install-and-run#start-the-typedb-server). Check that it's listening to address: `0.0.0.0:1729`.
3. Launch `typedb-examples/commerce/bookstore/python/load_data.py` python script. It will load the bookstore schema and data into the DB.
4. Launch `typedb-examples/commerce/bookstore/python/requests.py` python script and follow the instructions to explore example functions and data.
5. Use this simple example to learn the basics of using TypeDB! Explore source codes of the both python scripts and TypeDB Studio to explore DB schema and content.

## How it works

### Files

> This example located in the `typedb/commerce/bookstore/` directory and consists of the following main files:

Reviewer: ```suggestion
This example is located in the `typedb/commerce/bookstore/` directory and consists of the following main files:
```
Scan the rest of the README for typos / little errors.

Author: These little errors - are hard to find by the author. Thank you for your help, i'll try to proof read it again.

* [TypeDB](https://docs.vaticle.com/docs/running-typedb/install-and-run) v2.14.1+
* Python v.3.9+
  * `typedb.client` — [Python client](https://docs.vaticle.com/docs/client-api/python) for TypeDB

>   * Common Python libraries: csv, argparse, uuid, random

Reviewer: Very good to include this. This - as well as the addition of a `BUILD` file - reminds me that every other example, currently, is a Bazel project. We most likely do not want to diverge from this.

Converting a Python project to a Bazel project is very easy. See https://github.com/vaticle/typedb-examples/tree/master/telecom/phone_calls/python for inspiration. Basically we just need to add some `py_binary` and `py_library` targets to our `BUILD` file.

- `py_binary` represents an entry point; a Python script that the user is expected to run.
- `py_library` represents a library of Python functions; a Python script that is expected to be only referenced through "import" statements, and not run as a standalone binary.

So for the current project, `config` and `loaders` would be libraries, and `load_data` and `requests` would be binaries.

You can pretty much copy the `py_library` and `py_binary` syntax from this `BUILD` file: https://github.com/vaticle/typedb-examples/blob/master/telecom/phone_calls/python/BUILD

The other thing to do is to specify these requirements (csv, argparse, uuid, random) in a `requirements.txt` file. I believe `bazel run //commerce/bookstore:queries` will then automatically download these dependencies and just run - the "one-click run" that Bazel gives us.

Author: I have tried fixing this. But so far was unsuccessful. I will try again, maybe do some quickstart with Bazel. So far our Build files seem easy to understand but are still very tricky to make it work somehow.
So far only the checkstyle test working and I have already updated the files to pass this test. I'm considering adding unit tests to the example as it seems an easier task than making Bazel work so far.

Author: I have managed to make this work https://github.com/vaticle/typedb-examples/pull/143/commits/240df54e638d9a14bbde669a8d4190cad182be6f

Reviewer: Awesome!

You've built the project to work when built using Python and `pip` directly, however all of our projects including example projects are "Bazel first". I would therefore propose the following:

- Split up the Prerequisites/Quickstart sections into two higher-level blocks: `Quickstart (Bazel)` and `Quickstart (with custom Python installation)`.
- Within each `Quickstart`, create two lower-level sections: `Prerequisites` and `Run`. The non-Bazel guidelines are already pretty generic: they could work whether the user prefers to use Pip, Anaconda or literally download packages by hand. So all we need to do is add the Bazel section, which you've actually proposed already.

The principal advantage of this approach is we maintain consistency with our other examples by advocating Bazel and explaining how to run them using Bazel, while also providing clear guidelines to get those started who don't wish to install a new build system into their environment.

Author: Why would we need to do that? It's just an extra step. So the onboarded engineer would need to learn TypeDB and Bazel. That seems counter-productive to me. Like the very opposite of what we should do in the example.

Adding sophistication to example documentation is very undesirable.

Reviewer: This conversation has not aged well. Marking as resolved since no more Bazel usage!
